app/main/user-agent-service.js

```diff
@@ -0,0 +1,220 @@
+
+ Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ this file except in compliance with the License. You may obtain a copy of the
+ License at http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software distributed
+ under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ specific language governing permissions and limitations under the License.
+ */
+
+/* eslint no-console: 0 */
+
+import express from 'express';
+import expressValidator from 'express-validator';
+import bodyParser from 'body-parser';
+import morgan from 'morgan';
+import * as Task from 'co-task';
+
+import { ProfileStorage, StarOp } from '../services/storage';
+
+const PORT = 9090;
+
+function configure(app: any, storage: ProfileStorage) {
+  app.use(morgan('combined'));
+
+  // These need to be register before our routes.
+  app.use(bodyParser.json());
+  app.use(expressValidator()); // Keep this immediately after express.bodyParser().
+
+  /**
+   * Catch errors from the given generator function and forward them to `next` for handling.
+   *
+   * @param genfun: function*(req, res, next)
+   * @returns {function(req, res, next): (Promise.<T>)}
+   */
+  function wrap(genfun) {
+    return (...args) => Task.async(genfun)(...args).catch(args[2]); // next = args[2];
+  }
+
+  app.post('/session/start', wrap(function* (req, res) {
+    req.checkBody('scope').isInt().notEmpty();
+    req.checkBody('ancestor').isInt().optional();
+
+    const errors = req.validationErrors();
+    if (errors) {
+      res.status(401).json(errors);
+      return;
+    }
+
+    const { scope, ancestor } = req.body;
+    const session = yield storage.startSession(scope, ancestor);
+    res.json({ session });
+  }));
+
+  app.post('/session/end', wrap(function* (req, res) {
+    req.checkBody('session').isInt().notEmpty();
+
+    const errors = req.validationErrors();
+    if (errors) {
+      res.status(401).json(errors);
+      return;
+    }
+
+    const { session } = req.body;
+    yield storage.endSession(session);
+    res.json();
+  }));
+
+  app.post('/visits', wrap(function* (req, res) {
+    req.checkBody('url').notEmpty();
+    req.checkBody('title').optional();
+    req.checkBody('session').isInt().notEmpty();
+
+    const errors = req.validationErrors();
+    if (errors) {
+      res.status(401).json(errors);
+      return;
+    }
+
+    // TODO: include visit types.
+    const { url, session, title } = req.body;
+    yield storage.visit(url, session, title);
+    res.json();
+  }));
+
+  app.get('/visits', wrap(function* (req, res) {
+    req.checkQuery('q').notEmpty();
+
+    const errors = req.validationErrors();
+    if (errors) {
+      res.status(401).json(errors);
+      return;
+    }
+
+    const { q } = req.query;
+    const results = yield storage.visitedMatches(q);
+    res.json({ results });
```

Comment:
We're probably going to want the richer types specified in `AwesomebarResult` in my current branch, or at least something more than URLs… unless we only want to support URL completion here.